e2e/quit-on-close.e2e.spec.ts

```diff
@@ -0,0 +1,59 @@
+import path from 'path';
+import util from 'util';
+
+import { test, expect } from '@playwright/test';
+import { ElectronApplication, _electron, Page } from 'playwright';
+
+import { createDefaultSettings, packageLogs, reportAsset } from './utils/TestUtils';
+
+const waitTime = 1000;
+const sleep = util.promisify(setTimeout);
+
+async function startRancherDesktop(): Promise<[ElectronApplication, Page]> {
+  const electronApp = await _electron.launch({
+    args: [
+      path.join(__dirname, '../'),
+      '--disable-gpu',
+      '--whitelisted-ips=',
+      // See pkg/rancher-desktop/utils/commandLine.ts before changing the next item as the final option.
+      '--disable-dev-shm-usage',
+      '--no-modal-dialogs',
+    ],
+    env: {
+      ...process.env,
+      RD_LOGS_DIR: reportAsset(__filename, 'log'),
+    },
+  });
+  const page = await electronApp.firstWindow();
+  return [electronApp, page];
+}
+
+/**
+ * Using test.describe.serial make the test execute step by step, as described on each `test()` order
+ * Playwright executes test in parallel by default and it will not work for our app backend loading process.
+ * */
+test.describe.serial('quitOnClose setting', () => {
+  test.afterEach(async() => {
+    await packageLogs(__filename);
+  });
+
+  test('should quit when quitOnClose is true and window is closed', async() => {
+    createDefaultSettings({ window: { quitOnClose: true } });
+    const [electronApp, page] = await startRancherDesktop();
+    await sleep(waitTime);
+    const browserWindowHandle = await electronApp.browserWindow(page);
+    browserWindowHandle.evaluate((browserWindow: Electron.BrowserWindow) => browserWindow.close());
+    await sleep(waitTime);
```

Comment:
I've tried to get something like this working, but haven't been able to. The way I've tried is to listen to the `quit` event on `Electron.app`, and set a variable when the event fires:
```
  test('should quit when quitOnClose is true and window is closed', async() => {
    createDefaultSettings({ window: { quitOnClose: true } });
    const {electronApp, page} = await startRancherDesktop(__filename);
    const browserWindowHandle = await electronApp.browserWindow(page);

    let hasQuit = false;
    electronApp.evaluate(({app}) => {
      app.on('quit', () => {
        console.log('quit event occurred')
        hasQuit = true;
      });
    });

    browserWindowHandle.evaluate((browserWindow: Electron.BrowserWindow) => browserWindow.close());

    const sleepTime = 5000;
    const iterCount = 6;
    for (let i = 0; i < iterCount; i++) {
      console.log(`hasQuit: ${ hasQuit }`)
      if (hasQuit) {
        return;
      }
      await sleep(sleepTime)
    }
    // fail the test if RD has not quit by now
    expect(true).toBe(false);
  });
```
But it doesn't seem to work. I'm not very confident with playwright though. Maybe you can see something I'm doing wrong? Or maybe you're away of another way of doing it? I can't think of anything else...

Reply:
I believe `.evaluate` works something like serializing the function to text, and `eval()` it on the other side. So in your case, `hasQuit` _isn't_ getting a closure, so it's never set to true on the test side.

In my case, I do _everything_ within `evaluate()`, and it only returns a `Promise<boolean>` which the documentation says it can.